# 4.4 Data Wrangling & Subsetting 

## 01. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

## 02. Import Data

In [2]:
path = r'C:\Users\ThinkPad X250\Documents\Instacart Basket Analysis'

In [3]:
df_ords = pd.read_csv(r'C:\Users\ThinkPad X250\Documents\Instacart Basket Analysis\02 Data\Original Data\orders.csv', index_col = False)

In [4]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [5]:
df_prods = pd.read_csv(r'C:\Users\ThinkPad X250\Documents\Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

In [6]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [7]:
df_dep = pd.read_csv(r'C:\Users\ThinkPad X250\Documents\Instacart Basket Analysis\02 Data\Original Data\departments.csv', index_col = False)

In [8]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

## 03. Task - Data Wrangling Procedures

### 03.1 Orders (df_ords)

##### + To remove columns

In [9]:
df_ords = df_ords.drop(columns = ['eval_set'])

In [10]:
df_ords.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

##### + To find missing value

In [11]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

##### + To rename (obscure or unintuitive name)

In [12]:
df_ords.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [13]:
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

##### + To change data types from numbers to string columns 

In [14]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

##### + To check the new data type of a single column

In [15]:
df_ords['order_id'].dtype

dtype('O')

### 03.2 Departments (df_dep)

##### + To change data from wide format to long format

In [16]:
# department data

df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

##### + To create a new df

In [17]:
df_dep_t = df_dep.T

In [18]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

##### + To add an index

In [19]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

##### + To create a new header 

In [20]:
# 1. Create a new header

new_header = df_dep_t.iloc[0]

In [21]:
new_header

0    department
Name: department_id, dtype: object

In [22]:
# 2. Remove the first row of the dataframe

df_dep_t_new = df_dep_t[1:]

In [23]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [24]:
# 3. Add new header

df_dep_t_new.columns = new_header 

In [25]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

##### + To create a data dictionary

In [26]:
data_dict = df_dep_t_new.to_dict('index')

In [27]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

##### + To get 

In [28]:
print(data_dict.get('19'))

{'department': 'snacks'}


##### + To create a subset

In [29]:
df_snacks = df_prods[df_prods['department_id']==19]

In [30]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

## 03. Task (2-13)

##### 2. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [31]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [32]:
df_ords['user_id'].dtype

dtype('O')

##### 3.	Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the data frame. 

In [33]:
df_ords.rename(columns = {'orders_day_of_week' : 'order_day_of_week'})

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0        2539329       1             1                  2                  8   
1        2398795       1             2                  3                  7   
2         473747       1             3                  3                 12   
3        2254736       1             4                  4                  7   
4         431534       1             5                  4                 15   
...          ...     ...           ...                ...                ...   
3421078  2266710  206209            10                  5                 18   
3421079  1854736  206209            11                  4                 10   
3421080   626363  206209            12                  1                 12   
3421081  2977660  206209            13                  1                 12   
3421082   272231  206209            14                  6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

##### 4.	Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.
##### busiest hour for placing orders = 10 am

In [34]:
df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

##### 5.	Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [35]:
print(data_dict.get('4'))

{'department': 'produce'}


##### 6.	The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [36]:
df_brkfast = df_prods[df_prods['department_id']==14]

In [37]:
df_brkfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

##### 7.	They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [38]:
df_alco =  df_prods[df_prods['department_id']==5]

In [39]:
df_alco 

product_id                               product_name  aisle_id  \
51             52                        Mirabelle Brut Rose       134   
118           119                     Chardonnay Paso Robles        62   
149           150                                  Brut Rosé       134   
233           234                          Tennessee Whiskey       124   
248           249             Pinot Grigio, California, 2010        62   
...           ...                                        ...       ...   
49548       49544                    Cabernet Sauvignon Wine        28   
49566       49562              Blanc De Noirs Sparkling Wine       134   
49610       49606                        Organic Natural Red        28   
49665       49661                                      Porto       134   
49688       49684  Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
51                 5    14.4  
118                5     5.5  
149                5    12.9  
233                5     3.1  
248                5     2.7  
...              ...     ...  
49548              5     9.8  
49566              5     5.7  
49610              5     6.8  
49665              5     8.2  
49688              5     5.3  

[1056 rows x 5 columns]

In [40]:
df_bev =  df_prods[df_prods['department_id']==7]

In [41]:
df_bev

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
19             20  Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
...           ...                                             ...       ...   
49659       49655                                     Apple Cider        98   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
19                 7     6.0  
...              ...     ...  
49659              7    10.7  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  

[4365 rows x 5 columns]

In [42]:
df_deli =  df_prods[df_prods['department_id']==20]

In [43]:
df_deli

product_id                                   product_name  aisle_id  \
48             49  Vegetarian Grain Meat Sausages Italian - 4 CT        14   
61             62        Premium Deli Oven Roasted Turkey Breast        96   
73             74                Artisan Chick'n & Apple Sausage        14   
84             85                           Soppressata Piccante        96   
108           109                         Grape Leaf Hummus Wrap        13   
...           ...                                            ...       ...   
49558       49554                          Roasted Garlic Hommus        67   
49564       49560    Selects Natural Slow Roasted Chicken Breast        96   
49585       49581                   Pinto Bean and Cheese Pupusa        13   
49609       49605                     Classic Hummus Family Size        67   
49642       49638           Organic Superfirm Vacuum Packed Tofu        14   

       department_id  prices  
48                20    10.1  
61                20    14.6  
73                20     3.0  
84                20     8.9  
108               20     7.1  
...              ...     ...  
49558             20    14.8  
49564             20    14.5  
49585             20    10.5  
49609             20     3.5  
49642             20     9.0  

[1322 rows x 5 columns]

In [44]:
df_meat_sea =  df_prods[df_prods['department_id']==12]

In [45]:
df_meat_sea

product_id                                       product_name  \
16             17                                  Rendered Duck Fat   
22             23                             Organic Turkey Burgers   
34             35     Italian Herb Porcini Mushrooms Chicken Sausage   
39             40  Beef Hot Links Beef Smoked Sausage With Chile ...   
83             84                                         Lamb Shank   
...           ...                                                ...   
49425       49421                                   Salame, Peppered   
49440       49436                              Imitation Crab Flakes   
49509       49505                                Hot Italian Sausage   
49655       49651                                       Beef Brisket   
49674       49670                                Beef Summer Sausage   

       aisle_id  department_id  prices  
16           35             12    17.1  
22           49             12     8.2  
34          106             12    15.1  
39          106             12    22.5  
83            7             12    24.3  
...         ...            ...     ...  
49425       106             12    12.2  
49440        15             12    23.5  
49509       106             12    15.8  
49655       122             12    20.7  
49674       106             12    19.2  

[907 rows x 5 columns]

In [46]:
df_dinner = df_prods.loc[df_prods['department_id'].isin([5,7,20,12])]

In [47]:
df_dinner

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

##### 8.	It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?
##### -- 7650 rows x 5 columns

##### 9.	Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [48]:
df_ords_user1 = df_ords[df_ords['user_id']=='1']

In [49]:
df_ords_user1

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   
5   3367565       1             6                  2                  7   
6    550135       1             7                  1                  9   
7   3108588       1             8                  1                 14   
8   2295261       1             9                  1                 16   
9   2550362       1            10                  4                  8   
10  1187899       1            11                  4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

##### 11.	Check the organization and structure of your notebook. Be sure to include section headings and code comments.

In [50]:
df_ords_user1.describe()

order_number  order_day_of_week  order_hour_of_day  \
count     11.000000          11.000000          11.000000   
mean       6.000000           2.636364          10.090909   
std        3.316625           1.286291           3.477198   
min        1.000000           1.000000           7.000000   
25%        3.500000           1.500000           7.500000   
50%        6.000000           3.000000           8.000000   
75%        8.500000           4.000000          13.000000   
max       11.000000           4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

##### 12.	Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [51]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

##### 13.	Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [52]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))